### PR TITLE
OCD model migration

### PR DESCRIPTION
**Pull request by fgregg**
This PR changes django-councilmatic to use opencivicdata models.

In concert with https://github.com/datamade/chi-councilmatic/pull/252, these code changes can largely reproduce the Chicago Councilmatic site.

TODO:

- [x] rescue some commits from an old computer
- [x] clean up migrations
- [x] catch up with master
- [x] get tests passing
- [ ] get changes up on staging site
- [x] mayor is not member of council so his role does not show in list of sponsors http://127.0.0.1:8000/legislation/o2018-8860/
- [ ] ~add order_by to meta class on EventAgendaItem https://github.com/opencivicdata/python-opencivicdata/blob/master/opencivicdata/legislative/models/event.py~
- [ ] import data command for shapes (maybe this should use fixtures?)
- [ ] ~figure out how we want to handle updating the index~
- [x] document use of proxymodels
- [x] remove update_data management command.

### CODE REVIEW

**Comment by fgregg** [comment]
@hancush this is ready for me to indoctrinate you.

**Comment by hancush** [comment]
logging to return to later: some memberships don't have end dates in the ocd api, e.g., rahm https://ocd.datamade.us/ocd-person/f649753d-081d-4f22-8dcf-3af71de0e6ca/

[in django-councilmatic](https://github.com/datamade/django-councilmatic/blob/18b8286c94e37323b9220b32ebc9261c26435b96/councilmatic_core/models.py#L640), we store null values as null, but ocd [stores null values as empty strings](https://github.com/opencivicdata/python-opencivicdata/blob/e2d12dc81d34361c23c3ce0d5eb7c1a28ec8ed57/opencivicdata/core/models/people_orgs.py#L555).

this causes things to break when we perform operations that assume a certain type, e.g.,

https://github.com/datamade/django-councilmatic/blob/daad426bc60b585464eb65535e78b930e6bbe707/councilmatic_core/models.py#L250-L252

this may come to bear on other fields, too.

File: councilmatic_core/management/commands/update_headshots.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+import requests
+
+class Command(BaseCommand):
+    help = 'Attach headshots to councilmembers'
+
+    def handle(self, *args, **opetions):
+        for person in OCDPerson.objects.exclude(image=''):
+            councilmatic_person = person.councilmatic_person
+            filename = 'ocd-person-' + councilmatic_person.slug + '.jpg'
```
Comment (hancush) [inline councilmatic_core/management/commands/update_headshots.py]: i added this prefix so you can easily exclude these files in a gitignore.

File: councilmatic_core/models.py [hunk 2]
```diff
@@ -1,96 +1,110 @@
...
-    raise ImproperlyConfigured(
-        'You must define a OCD_CITY_COUNCIL_ID or OCD_CITY_COUNCIL_NAME in settings.py')
+from proxy_overrides.related import ProxyForeignKey
+import opencivicdata.legislative.models
+import opencivicdata.core.models
 
-if not hasattr(settings, 'CITY_COUNCIL_NAME'):
-    raise ImproperlyConfigured(
-        'You must define a CITY_COUNCIL_NAME in settings.py')
+static_storage = FileSystemStorage(location=os.path.join(settings.BASE_DIR, settings.APP_NAME, 'static'), base_url='/')
```
Comment (hancush) [inline councilmatic_core/models.py]: i could only get the downloaded headshots to work if they were saved to the app static directory, then moved to the static root via collectstatic.

File: councilmatic_core/models.py [hunk 3]
```diff
@@ -396,158 +584,46 @@ def updated_bills_since(cls, date_cutoff):
...
-
-            return link_fmt.format(link_path, self.name)
-
-        return self.name
+class BillSponsorship(opencivicdata.legislative.models.BillSponsorship):
+    class Meta:
+        proxy = True
 
+    bill = ProxyForeignKey(Bill, related_name='sponsorships', on_delete=models.CASCADE)
+    person = ProxyForeignKey(Person, null=True, on_delete=models.SET_NULL)
```
Comment (hancush) [inline councilmatic_core/models.py]: interesting choice to make this nullable. in what instances are sponsorships not associated with a person? here are the sponsorships without associated people on my local machine:

```
| id                                   | name                       | entity_type | primary | classification | bill_id                                       | organization_id | person_id |
| ------------------------------------ | -------------------------- | ----------- | ------- | -------------- | --------------------------------------------- | --------------- | --------- |
| f0f9e1dc-108c-44b6-9209-78465bbfa023 | Lightfoot, Lori E.         | person      |    True | Primary        | ocd-bill/5b6563dd-f022-4584-9e87-4ef662575378 |                 |           |
| 3d36f500-083c-4cda-866a-31278d277e80 | Lightfoot, Lori E.         | person      |    True | Primary        | ocd-bill/4a712ecd-da48-43e8-8158-a1170bc5bec9 |                 |           |
| a7f9dc13-30f2-41ef-a581-fca2621c735b | La Spata, Daniel           | person      |    True | Primary        | ocd-bill/4bd88d59-e39e-468b-9643-5e1b59c23674 |                 |           |
| 070f437a-26ee-4a14-b9cd-1e7de83fdc59 | Coleman, Stephanie D.      | person      |   False | Regular        | ocd-bill/4bd88d59-e39e-468b-9643-5e1b59c23674 |                 |           |
| 4eea1d02-e6db-47d5-89b0-ca79f2eab604 | Taylor, Jeanette B.        | person      |   False | Regular        | ocd-bill/4bd88d59-e39e-468b-9643-5e1b59c23674 |                 |           |
| b90ca306-82a0-4aa1-8726-3f18a150c452 | Rodriguez, Michael D.      | person      |   False | Regular        | ocd-bill/4bd88d59-e39e-468b-9643-5e1b59c23674 |                 |           |
| 6fa576fb-b99b-4025-9d88-bf929d0b7454 | Byron Sigcho-Lopez         | person      |   False | Regular        | ocd-bill/4bd88d59-e39e-468b-9643-5e1b59c23674 |                 |           |
| 09bceabf-e54e-4e45-ba47-644ae6482e41 | Cardona, Jr., Felix        | person      |   False | Regular        | ocd-bill/4bd88d59-e39e-468b-9643-5e1b59c23674 |                 |           |
| 055e6e86-0156-4f44-9ba3-48df89c94aac | Rodriguez Sanchez, Rossana | person      |   False | Regular        | ocd-bill/4bd88d59-e39e-468b-9643-5e1b59c23674 |                 |           |
| 1d7367aa-a4da-4e54-ae11-c7997c31023a | Nugent, Samantha           | person      |   False | Regular        | ocd-bill/4bd88d59-e39e-468b-9643-5e1b59c23674 |                 |           |
| d9d44823-9f3f-4ecd-a1fa-07b50f0f8361 | Vasquez, Jr., Andres       | person      |   False | Regular        | ocd-bill/4bd88d59-e39e-468b-9643-5e1b59c23674 |                 |           |
| 2f09b027-49bc-491a-a704-4dde4b82f628 | Gardiner, James M.         | person      |   False | Regular        | ocd-bill/4bd88d59-e39e-468b-9643-5e1b59c23674 |                 |           |
| eb07d993-757e-41b7-98e8-fa59df081f22 | Martin, Matthew J.         | person      |   False | Regular        | ocd-bill/4bd88d59-e39e-468b-9643-5e1b59c23674 |                 |           |
| 70c011f1-ab3e-4742-9ef0-8ad42a6fbe01 | Hadden, Maria E.           | person      |   False | Regular        | ocd-bill/4bd88d59-e39e-468b-9643-5e1b59c23674 |                 |           |
| 8f8bcc77-bf07-480c-9833-1bf029e21a0d | Lightfoot, Lori E.         | person      |    True | Primary        | ocd-bill/0866399e-385d-4030-8158-fa4de682e614 |                 |           |
| 872a5b35-5e33-42db-9e07-e4cc23b8a489 | Lightfoot, Lori E.         | person      |    True | Primary        | ocd-bill/cecaf55f-b8a5-4fff-9221-330a73a27ad6 |                 |           |
| d97ced7c-728d-48a5-93d4-eb809945497b | Lightfoot, Lori E.         | person      |    True | Primary        | ocd-bill/3317b40f-9db1-4465-82cb-b70240177583 |                 |           |
| d6eea56a-7205-4f17-963e-8702a49bbec3 | Lightfoot, Lori E.         | person      |    True | Primary        | ocd-bill/56048709-7b05-4526-b270-5c6b0bf2d840 |                 |           |
| 0c8be0f5-770a-4b5b-bac8-91799e22b759 | Lightfoot, Lori E.         | person      |   False | Regular        | ocd-bill/e4548dee-9a08-4d62-a7e4-604358e4c6bf |                 |           |
| 4a11f867-be96-4eef-acce-b79406a0438a | Lightfoot, Lori E.         | person      |    True | Primary        | ocd-bill/21079f96-5361-4d6a-b7a4-61fa663746f9 |                 |           |
```

File: councilmatic_core/templates/councilmatic_core/legislation.html [hunk 4]
```diff
@@ -104,15 +104,24 @@ <h3 class="no-pad-bottom"><i class='fa fa-fw fa-users'></i> Sponsors</h3>
                         </thead>
                         <tbody>
                             {% for s in legislation.sponsorships.all %}
-                                {% if s.is_primary %}
+                                {% if s.primary %}
                                 <tr>
                                     <td>
                                         <div class="thumbnail-square">
-                                            <img src='{{s.person.headshot_url}}' alt='{{s.person.name}}' title='{{s.person.name}}' class='img-responsive img-thumbnail' />
+                                            {% if s.person %}
```
Comment (hancush) [inline councilmatic_core/templates/councilmatic_core/legislation.html]: may want to update this, depending on how we want to handle sponsorships without associated person objects.

File: councilmatic_core/management/commands/make_fixtures.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import os
+
+import django
+from django.conf import settings
+from django.core.management import call_command
+from django.core.management.base import BaseCommand
+from django.db import connection
+
+
+class Command(BaseCommand):
```
Comment (hancush) [inline councilmatic_core/management/commands/make_fixtures.py]: it was a massive pita to refresh the test fixtures. this is a convenience command leveraging [my fork of django-fixture-magic](https://github.com/hancush/django-fixture-magic/tree/fix-merge) to dump the core models, and their relations. i use this, rather than dump data, because it allows you to dump a subset of your database and automatically grabs related items. dumping the full instance db resulted in an enormous fixture file that exponentially increased test runtime.

File: councilmatic_core/management/commands/refresh_pic.py [hunk 6]
```diff
@@ -56,24 +56,21 @@ def _get_urls(self):
         with connection.cursor() as cursor:
 
             query = '''
-                SELECT DISTINCT url from 
-                councilmatic_core_billdocument as b_doc
-                JOIN change_bill 
-                ON change_bill.ocd_id=b_doc.bill_id
-                WHERE document_type='V'
```
Comment (hancush) [inline councilmatic_core/management/commands/refresh_pic.py]: note that ocd does not include a good way to store the A/V values indicating whether a document is a version. we use this to filter this query and one in convert_attachment_text. i'm not sure what the consequences of omitting them will be.

Comment (hancush) [reply councilmatic_core/management/commands/refresh_pic.py]: actually nvm

**Comment by hancush** [comment]
logging for myself: chicago doesn't have bill documents! 

```
select * from opencivicdata_billdocument as doc join opencivicdata_bill as bill on doc.bill_id = bill.id join opencivicdata_organization as org on bill.from_organization_id = org.id where org.jurisdiction_id = 'ocd-jurisdiction/country:us/state:il/place:chicago/government';
 id | note | date | bill_id | created_at | updated_at | extras | id | identifier | title | classification | subject | legislative_session_id | from_organization_id | locked_fields | created_at | updated_at | extras | id | name | image | classification | founding_date | dissolution_date | jurisdiction_id | parent_id | locked_fields
----+------+------+---------+------------+------------+--------+----+------------+-------+----------------+---------+------------------------+----------------------+---------------+------------+------------+--------+----+------+-------+----------------+---------------+------------------+-----------------+-----------+---------------
(0 rows)
```

**Comment by hancush** [comment]
@fgregg making good progress here.

can you say more about:

- add order_by to meta class on EventAgendaItem
- figure out how we want to handle updating the index

?

**Comment by hancush** [comment]
re: `add order_by to meta class on EventAgendaItem`, there's already `clean_agenda_items` which orders and dedupes agenda items.

https://github.com/datamade/django-councilmatic/blob/8d64575cf93470d13db99db9f4d42afd3a67cc5c/councilmatic_core/models.py#L711-L721

**Comment by fgregg** [comment]
if you add ordering to the model class, it will, by default return querysets in that order. This is just an opportunity to clean up some code, and not required for this migration.

https://docs.djangoproject.com/en/2.2/ref/models/options/#ordering

**Comment by hancush** [comment]
# Change log

## Description

tl;dr - Remove the need for extensive and error-prone ETL by basing the Councilmatic app off Open Civic Data models, i.e., a jurisdiction-specific database fed by the scrapers, rather than maintaining a remote set of similar, Councilmatic-specific models.

There is a Python package containing Django versions of the OCD models. The relevant ones live in [`opencivicdata.core.models`](https://github.com/opencivicdata/python-opencivicdata/tree/master/opencivicdata/core) and [`opencivicdata.legislative.models`](https://github.com/opencivicdata/python-opencivicdata/tree/master/opencivicdata/legislative/models).

Further reading:

- https://github.com/datamade/devops/issues/59
- https://github.com/datamade/django-councilmatic/issues/241

## Approach

### Extend Open Civic Data models

#### Subclassing

Leverage [multi-table inheritance](https://docs.djangoproject.com/en/2.2/topics/db/models/#multi-table-inheritance) to add additional fields to OCD models. The primary use case in the refactor is adding slugs to first-class models – Person, Event, Bill, and Organization - and adding a headshot to Person.

Since the scrapers create OCD objects, [we added signals](https://github.com/datamade/django-councilmatic/pull/240/files#diff-97cdca8c3c4b594b1991875f343b7db5) to each of the subclassed models to create the related Councilmatic model on save. [Read more on Django signals. &raquo;](https://docs.djangoproject.com/en/2.2/topics/signals/)

We have also added a management command to import headshot files.

#### Proxying

We make extensive use of proxy models to add custom managers and additional properties and methods to classes for use in Councilmatic code. This is an issue because the proxied models return OCD objects, not Councilmatic objects. To get around that, we used the hacky but effective `django-proxy-overrides` library to force models to return related objects as Councilmatic models, as needed.

To customize the model of related objects, proxy a model like [`opencivicdata.legislative.models.BillAction`](https://github.com/opencivicdata/python-opencivicdata/blob/e2d12dc81d34361c23c3ce0d5eb7c1a28ec8ed57/opencivicdata/legislative/models/bill.py#L82-L102) -> [`councilmatic_core.models.BillAction`](https://github.com/datamade/django-councilmatic/blob/00b08ef26ffc5e6cd3bda6ad72c04b2ca49460ad/councilmatic_core/models.py#L613-L626), and override one or more related object attributes with a `ProxyForeignKey`.

A brief summary of alternative approaches and why we ultimately did not choose to use them [can be found here](https://docs.google.com/document/d/1VkasSM6yVrone8gEHFl4eEF3o7JVZhttfL7L-mRLx4E/edit?usp=sharing).

#### Related changes

- Update attribute references
- Handle changes in data type, e.g., cast date strings to datetime objects
- Refactor use of obsolete raw SQL, e.g., management commands

### Upgrade to Django 2

- Update some module import paths
- staticfiles -> static template tag
- Build URLs via static template tag

### Repair and extend the tests

- Centralize fixtures and add a management command to regenerate them from a local instance DB
- Align individual fixtures with the model change

## Remaining changes

- Update the councilmatic-starter-template to use the refactored release
  - To that end, see https://github.com/datamade/chi-councilmatic/pull/252#issuecomment-494466071 for startup documentation
- Add a management command to import shapes
- Repair the notifications app

File: councilmatic_core/models.py (hunk 2 above)
Comment (hancush) [reply councilmatic_core/models.py]: added a call to collectstatic to the end of this command

File: councilmatic_core/models.py (hunk 3 above)
Comment (fgregg) [reply councilmatic_core/models.py]: An organization can be a sponsor.

**Review by fgregg — COMMENTED** [review]
overall looks great! there are some thing would be nice to change. I can't mark this as a request change.

File: councilmatic_core/management/commands/convert_attachment_text.py [hunk 7]
```diff
@@ -38,41 +38,30 @@ def handle(self, *args, **options):
...
+            qs = BillDocumentLink.objects.filter(is_null & is_file)
+        else:
+            qs = BillDocumentLink.objects.filter(is_null & is_file & after_max_update)
 
-            yield from result
+        for item in qs:
+            yield item.url, item.document.id
 
     def convert_document_to_plaintext(self):
+        import textract
```
[the remark this answers is not in the document]
Comment (hancush) [reply councilmatic_core/management/commands/convert_attachment_text.py]: so you can run the tests w/o needing to install textract, which is a pita. https://textract.readthedocs.io/en/stable/installation.html / https://github.com/datamade/django-councilmatic/pull/193#issuecomment-378748419

[the remark this answers is not in the document]
Comment (fgregg) [reply councilmatic_core/management/commands/convert_attachment_text.py]: please add a comment to that effect.

File: councilmatic_core/haystack_indexes.py [hunk 8]
```diff
@@ -58,16 +56,20 @@ def prepare_full_text(self, obj):
...
-        if not obj.last_action_date:
-            return datetime.now().replace(tzinfo=app_timezone) - timedelta(days=36500)
-        return obj.last_action_date
+        from datetime import timedelta
+        last_date = obj.last_action_date
+        if not last_date:
+            last_date = timezone.now() - timedelta(36500)
+
+        if last_date:
+            last_date = last_date - last_date.utcoffset()
```
[the remark this answers is not in the document]
Comment (hancush) [reply councilmatic_core/haystack_indexes.py]: i'll be honest with you, i have no idea

[the remark this answers is not in the document]
Comment (hancush) [reply councilmatic_core/haystack_indexes.py]: i think that an easier question to answer is: how should the index handle bills without a last action date? 

it seems like this is placing them way, way in the past. is this for the legislative session filter? if so, then it would have the effect of not returning bills with this funny date for any of the current legislative sessions. i think that we could achieve this by simply returning `None` if there is no `last_action_date`, no?

[the remark this answers is not in the document]
Comment (fgregg) [reply councilmatic_core/haystack_indexes.py]: i think that this has to do with https://github.com/datamade/chi-councilmatic/blob/master/chicago/search_indexes.py#L20-L26

but we should just handle that kind of changes in that code, and we override this method anyway. I think we can drop this method altogether.

[the remark this answers is not in the document]
Comment (hancush) [reply councilmatic_core/haystack_indexes.py]: done! rebuilt the chicago index to make sure it didn't break anything, and it looks ok to me!

**Review by fgregg — COMMENTED** [review]
approve

File: councilmatic_core/models.py (hunk 3 above)
Comment (fgregg) [reply councilmatic_core/models.py]: yes, that would be good.
